src/helm/common/object_spec.py

```diff
@@ -13,7 +13,10 @@ class ObjectSpec:
     args: Dict[str, Any]
 
     def __hash__(self):
-        return hash((self.class_name, tuple((k, self.args[k]) for k in sorted(self.args.keys()))))
+        t = tuple()
+        for k in sorted(self.args.keys()):
+            t = t + ((k, tuple(self.args[k])),)
```

Comment:
Unfortunately this now breaks on non-list values because `tuple(self.args[k])` doesn't work if `self.args[k]` is _not_ a list:

```
from helm.common.object_spec import ObjectSpec

o = ObjectSpec("myclass", {"myarg": 42})
hash(o)
```
Error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yifanmai/oss/helm/src/helm/common/object_spec.py", line 18, in __hash__
    t = t + ((k, tuple(self.args[k])),)
TypeError: 'int' object is not iterable
```

Could we do something like this instead?

```
    def __hash__(self):
        def get_arg_value(key: str) -> Any:
            value = self.args[key]
            # lists are not hashable, so convert them to tuples
            if isinstance(value, list):
                return tuple(value)
            return value

        args_tuple = tuple((k, get_arg_value(k)) for k in sorted(self.args.keys()))
        return hash((self.class_name, args_tuple))
```

Reply:
My bad, I should run few test.

The new code you suggest is again specific to List and premitive data type, which is good enough for our usecase, which will either have list of strings or a single string.

It works in these 2 datatype, however, it fails for dict.

Do you think this could have been a better solution for other data types as well for future proof? I was suggesting the same earlier too.


    import typing
    def __hash__(class_name, args):
        def get_arg_value(key: str):
            value = self.args[key]
            # lists are not hashable, so convert them to tuples
            if not isinstance(value, typing.Hashable):
                return value.__str__()
            return value
    
        args_tuple = tuple((k, get_arg_value(k)) for k in sorted(self.args.keys()))
        return hash((self.class_name, args_tuple))